File: python/src/find_anagrams.py
import hashlib 

def diff(first, last):
  first_arr = list(first)
  last_arr = list(last)
  for c in first:
    if c in last_arr:
      first_arr.remove(c)
      last_arr.remove(c)
  return "".join(first_arr)

def find_anagrams(phrase, candidate, words, secrets, result):
  filtered = [w for w in words if len(diff(w, phrase)) == 0]
  for w in filtered:
    new_candidate = [*candidate, w]
    new_words = [*filtered]
    new_words.remove(w)
    new_phrase = diff(phrase, w)
    if new_phrase == "":
      new_candidate_str = " ".join(new_candidate)
      digest = hashlib.md5(new_candidate_str.encode()).hexdigest()
      r = next(s for s in secrets if s == digest)
      if r:
        result.append(new_candidate_str)
    else:
      find_anagrams(new_phrase, new_candidate , new_words, secrets, result)


# import time
# import diff
# import find_anagrams

# phrase = "poultryoutwitsants"
# fo = open("words", "r")

# wordlist = [line.rstrip('\n') for line in fo]  # get words from file
# fo.close()
# # remove words include different chars
# wordlist = [w for w in wordlist if len(diff(w, phrase)) == 0]
# wordlist = list(dict.fromkeys(wordlist))  # remove dublicates
# wordlist = [w for w in wordlist if len(w) > 3]  # take words whose length

# start_time = time.time()
# result = []
# secrets = ["e4820b45d2277f3844eac66c903e84be", "23170acc097c24edb98fc5488ab033fe", ""]
# find_anagrams(phrase, [], wordlist, secrets, result)
# elapsed_time = time.time() - start_time
# print("Time Elapsed :{0:2f}". format(elapsed_time))

File: javascript/src/find_anagrams.js
import crypto from "crypto";

export function remove(array, element) {
    const index = array.indexOf(element);

    if (index !== -1) {
        array.splice(index, 1);
    }
}

export function find_diff(first, last) {
    let firstArr = [...first];
    let lastArr = [...last];
    for (let f of first) {
        if (lastArr.includes(f)) {
            remove(firstArr, f);
            remove(lastArr, f);
        }
    }
    return firstArr.join("");
}

export function find_anagrams(phrase, candidate, words, secrets, result) {
    let filtered = words.filter(w => !find_diff(w, phrase));
    for (let w of filtered) {
        let new_candidate = [...candidate, w];
        let new_words = [...words];
        remove(new_words, w);
        let new_phrase = find_diff(phrase, w);
        if (new_phrase === "") {
            let new_candidate_str = new_candidate.join(" ");
            let digest = crypto.createHash('md5').update(new_candidate_str).digest("hex");
            let r = secrets.find(x => x === digest);
            if (r) {
                result.push({ phrase: new_candidate_str, secret: digest });
            }
        }
        else {
            find_anagrams(new_phrase, new_candidate, new_words, secrets, result);
        }
    }
}

/*import fs from 'fs';
import { find_diff } from './find_diff';
import find_anagrams from './find_anagrams';
const phrase = "poultryoutwitsants";
const wordlist = [];

var lines = fs.readFileSync('words').toString().split("\n");
for(let line of lines){
    let word = line.replace("\r", "");
    if(word.length > 4 && !wordlist.includes(word) && !find_diff(word, phrase)){
        wordlist.push(word);
    }
}
console.time("dbsave");
let secrets = ["e4820b45d2277f3844eac66c903e84be", "23170acc097c24edb98fc5488ab033fe", ""]
let result = [];
find_anagrams(phrase, [], wordlist, secrets, result);
console.timeEnd("dbsave");*/